### DISCUSSION TITLE
Why build your own form library?

### DISCUSSION THREAD

**Discussion by brycefranzen** [mantinedev/mantine#5263] [Feedback]
I've used Mantine form pretty extensively on a large project of mine and have ran into a gazillion issues with it -- the entire form re-rendering on every input change, current form submission/validation state not being exposed, typescript types for field names (from the validation schema) are not enforced on `setFieldValue` and `getInputProps` function parameters, zod validation transformations don't occur prior to submit, useForm's `transform` doesn't allow changing the value's data type without typescript getting mad, the list goes on and on.

I've used [React Hook Form](https://react-hook-form.com/) on other larger projects and have never had any of these same issues, in fact everything just works.

It seems like a lot of extra work to engineer and maintain a form library when there is already an alternative that does everything your current library does and much, much more.  It has nearly the exact same base API, is significantly more performant, and is backed by an independent very large developer community.

Admittedly, yes, there are now ways to use React Hook Form with Mantine V7 using one of their 3rd-party bindings: https://github.com/aranlucas/react-hook-form-mantine

My suggestion is to ditch Mantine form and use React Hook Form instead.
Then, add these 3rd-party bindings into the Mantine form project.
This would allow users to utilize a significantly more powerful form library while seamlessly using Mantine components. 

 I'd genuinely like to know the reason why you built your own form library rather than using React Hook Form.

**Comment by rtivital** [comment]
I build the library to:
1. Have fun
2. Create a base for my personal projects and projects at work

Building a library from scratch is fun. Maintaining a third-party integration is not really fun. `@mantine/form` fully meets the requirements of all projects that I'm working on, so there is no reason to switch to another library and lose the most important fun part.

**Reply by brycefranzen** [reply]
fair enough. 😂

I absolutely love Mantine and all the hard work you've put into it doesn't go unnoticed.
Thanks!!!

**Reply by boggye** [reply]
> I build the library to:
> 
>     1. Have fun
> 
>     2. Create a base for my personal projects and projects at work
> 
> 
> Building a library from scratch is fun. Maintaining a third-party integration is not really fun. `@mantine/form` fully meets the requirements of all projects that I'm working on, so there is no reason to switch to another library and lose the most important fun part.

What if one day you cease to have fun? What will happen?

**Comment by aniforprez** [comment]
One of the biggest downsides to implementing a custom form library is that it doesn't play ball with native forms. Doing `FormData(form)` in a form with Mantine components gives you no information. This is breaking compatibility with a lot of things and adding custom handlers to create FormData before submitting is not ideal. And for some reason, names are not populated in fields inside the form which, again, breaks a lot of compatibility, especially with Remix who use native HTML features to do submissions. This, in my opinion, is breaking with the new ethos of v7 for Mantine acting as a headless UI library.

I would suggest not using the custom form hooks until something is done to maintain compatibility with native HTML forms.

**Reply by jamalsoueidan** [reply]
maybe use a third-party solution like @brycefranzen suggested. Embla and Tiptap are already being used in Mantine. I was considering abandoning the conform.guide form that I'm currently using, but I decided to research other people's experiences with mantine form. I've previously encountered issues with remix-validated-form and had to switch to conform.guide.

Now, the v1 upgrade of conform.guide has many breaking changes. I thought about using Mantine's form. Perhaps using conform.guide with Mantine could work; it's efficient and the library is small, and works with Remix/Next etc.

**Comment by flan02** [comment]
In Argentina a few group of people use mantine to stylize projects... I'm glad to be into that group... MANTINE ❤
